### ISSUE TITLE
No GitLab connection configured

### ISSUE THREAD

**Issue by AndreaCrotti** [jenkinsci/gitlab-plugin#450]
## Issue

Getting `No GitLab connection configured` even if there is only one connection and it's selected in the job.

The same gitlab connection is also selected in the job configuration, but then when I add the post build action to "publish the build status" I always get that error above.

Any idea of what the problem could be?
Thanks

### Context
- **Gitlab plugin version**: 1.4.3
- **Gitlab version**: 8.13.5 09cedb5
- **Jenkins version**: 2.19.3

### Problem description

_Describe your problem in a meaningful way_:
- The connection to gitlab is added to the global jenkins config and "test connection" works fine.
  The same gitlab connection is also selected in the job configuration.
   Proceeded then to add the post build action to publish commit status.

- I would expect a connection to gitlab or at least get an error which means it actually tried to connect. 
- instead it simply doesn't find the connection configuration somehow

**Comment by AndreaCrotti** [comment]
And by the way I wanted to give some more logs but even if I enabled the logging as in picture, nothing comes up from there.


![image](https://cloud.githubusercontent.com/assets/53640/20537379/db50f0a6-b0e4-11e6-8100-5585b5ee55b3.png)

**Comment by AndreaCrotti** [comment]
mm actually maybe it's because we are using CE and not EE after all,

    Got response from GitLab:
    URL: http://long_url..
    Status: 301 Moved Permanently
    Response headers: [
    Connection = [keep-alive],
    Content-Length = [178],
    Content-Type = [text/html],
    Date = [Thu, 24 Nov 2016 15:05:17 GMT],
    Location = [long_url],
    Server = [nginx]
    ]
    Response body: <html>
    <head><title>301 Moved Permanently</title></head>
    <body bgcolor="white">
    <center><h1>301 Moved Permanently</h1></center>
    <hr><center>nginx</center>
    </body>
    </html>

**Comment by AndreaCrotti** [comment]
One simpler question in case anyone knows.
From the docs of this project it looks like Gitlab CE should work, but from this:
https://gitlab.com/gitlab-org/gitlab-ce/issues/24217

maybe it's not, what's the reality?

**Comment by omehegan** [comment]
The plugin has historically always worked with both CE and EE. I'm not sure what this GitLab issue is referring to, but I don't think it's overall Jenkins support. @dblessing can you explain?

**Comment by AndreaCrotti** [comment]
By the way I tried on a more recent version of Jenkins (2.35) and it does seem to work, at least I see
```
[Set GitHub commit status (universal)] PENDING on repos [] (sha:d7cc6db) with context:something/TOXENV=py27
```
in the output.
However the build status is  still not set to the commit.
I tried manually with the API directly and it works, so the problem is definitively not on the gitlab site.

If it gives that message however it means in theory it's all correct right?
The other problem is that the state should have been FAILING instead of PENDING since the build clearly failed, but that's another story (not sure why is that either).

I wonder if I would be better off writing my own little script since it's so easy to use the API directly anwyay

**Comment by AndreaCrotti** [comment]
Ah sorry to be more clear it actually still give the same issue using for example
`Publish build status to GitLab commit (GitLab 8.1+ required)`

It's only by using:
`Set GitHub commit status (universal)`

that it actually looks like it's doing something even if it still doesn't work..

**Comment by AndreaCrotti** [comment]
Argh sorry, last two messages were relative to the Github plugin, which obviously has nothing to do with this.
Still not sure what the hell is reporting to given it's not even a gitbub plugin, but that's another story clearly..

So the problem is still the same "Gitlab connection not configured" which keeps on happening.

**Comment by omehegan** [comment]
@AndreaCrotti what type of job is this? Pipeline, Freestyle, something else?

**Comment by AndreaCrotti** [comment]
It's a quite simple Freestyle job, and there is one connection configured in the system configuration, which is selected as the gitlab connection in the job config.
The weird thing is that there is no way to deselect that either, not sure if that's a symptom that something is actually wrong..

In the end I already wrote all the code to handle it myself with a post job task and just doing the right calls with Python, but would be still nice to get it working with the plugin since we use it already anyway.

**Comment by omehegan** [comment]
Can you try creating a new job, configuring it the same way your current one is, and saving it? Look at the Jenkins log when you save. I'm wondering if it is throwing an error at that time, and causing the job configuration to be broken.

We have no other reports of this feature not working for Freestyle jobs.

**Comment by omehegan** [comment]
Can you also paste some screenshots of your GitLab config on the Jenkins global config page, and on the job config page?

**Comment by AndreaCrotti** [comment]
I'm sorry it was actually a MultiConfiguration job, using the `tox` builder that generates py27 and py35 jobs.

I tried with a freestyle job instead but it still fails even if in another way
```
Got response from GitLab:
URL: http://git.company.co.uk/api/v3/projects/company%2Fcompany-json/statuses/3159d2aaace7dfea7906eefc4e33b7d91e3000ed?state=success&context=jenkins&target_url=http%3A%2F%2Flocalhost%3A8080%2Fjob%2Fcompany-json-free%2F1%2F
Status: 301 Moved Permanently
Response headers: [
Connection = [keep-alive],
Content-Length = [178],
Content-Type = [text/html],
Date = [Tue, 13 Dec 2016 10:09:06 GMT],
Location = [https://git.company.co.uk:443/api/v3/projects/company%2Fcompany-json/statuses/3159d2aaace7dfea7906eefc4e33b7d91e3000ed?state=success&context=jenkins&target_url=http%3A%2F%2Flocalhost%3A8080%2Fjob%2Fcompany-json-free%2F1%2F],
Server = [nginx]
]
Response body: <html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx</center>
</body>
</html>63
```

Something funny with the "company%2Fcompany-json" I suppose.

But well generally I have most Multi-configuration jobs, I wonder exactly what happens if for example Py27 fails but Py35 doesn't for example, but that's another issue..

**Comment by omehegan** [comment]
Matrix/multi-configuration jobs are not supported by the plugin unfortunately, as noted in the documentation. Freestyle or Pipeline jobs should work.

The 301 error you get with a Freestyle job makes me think you might have a proxy in front of your GitLab server. Is that correct?

**Comment by AndreaCrotti** [comment]
Ah ok thanks @omehegan  sorry I missed that.
Well we probably have nginx in front of Gitlab, but even things like the requests python library or httpie still work without any extra arguments.
So gitlab-plugin doesn't follow the redirect?

**Comment by AndreaCrotti** [comment]
Ok for the records it was a silly mistake, simply the connection was configured with HTTP and not HTTPS, the redirect was simply our Nginx forcing HTTPS.
It all seems to work now, thanks for the help!
